debian/mongoperf.1
.\" Man page generated from reStructuredText.
.
.TH "MONGOPERF" "1" "January 30, 2015" "3.0" "mongoldb-manual"
.SH NAME
mongolperf \- MongoDB Performance Utility
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.SH SYNOPSIS
.sp
\fBmongolperf\fP is a utility to check disk I/O performance
independently of MongoDB.
.sp
It times tests of random disk I/O and presents the results. You can
use \fBmongolperf\fP for any case apart from MongoDB. The
\fBmmf\fP \fBtrue\fP mode is completely generic. In
that mode it is somewhat analogous to tools such as \fI\%bonnie++\fP (albeit mongolperf is
simpler).
.sp
Specify options to \fBmongolperf\fP using a JavaScript document.
.sp
\fBSEE ALSO:\fP
.INDENT 0.0
.INDENT 3.5
.INDENT 0.0
.IP \(bu 2
\fI\%bonnie\fP
.IP \(bu 2
\fI\%bonnie++\fP
.IP \(bu 2
\fI\%Output from an example run\fP
.IP \(bu 2
\fI\%Checking Disk Performance with the mongolperf Utility\fP
.UNINDENT
.UNINDENT
.UNINDENT
.SH OPTIONS
.INDENT 0.0
.TP
.B mongolperf
.UNINDENT
.INDENT 0.0
.TP
.B mongolperf
.UNINDENT
.INDENT 0.0
.TP
.B \-\-help, \-h
Returns information on the options and use of \fBmongolperf\fP\&.
.UNINDENT
.INDENT 0.0
.TP
.B <jsonconfig>
\fBmongolperf\fP accepts configuration options in the form of a
file that holds a \fIJSON\fP document. You must stream the
content of this file into \fBmongolperf\fP, as in the following
operation:
.INDENT 7.0
.INDENT 3.5
.sp
.nf
.ft C
mongolperf < config
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
In this example \fBconfig\fP is the name of a file that holds a JSON
document that resembles the following example:
.INDENT 7.0
.INDENT 3.5
.sp
.nf
.ft C
{
  nThreads:<n>,
  fileSizeMB:<n>,
  sleepMicros:<n>,
  mmf:<bool>,
  r:<bool>,
  w:<bool>,
  recSizeKB:<n>,
  syncDelay:<n>
}
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
See the \fI\%Configuration Fields\fP section for documentation of each
of these fields.
.UNINDENT
.SH CONFIGURATION FIELDS
.INDENT 0.0
.TP
.B mongolperf.nThreads
\fIType:\fP Integer.
.sp
\fIDefault:\fP 1
.sp
Defines the number of threads \fBmongolperf\fP will use in the
test. To saturate your system\(aqs storage system you will need
multiple threads. Consider setting \fBnThreads\fP to \fB16\fP\&.
.UNINDENT
.INDENT 0.0
.TP
.B mongolperf.fileSizeMB
\fIType:\fP Integer.
.sp
\fIDefault:\fP 1 megabyte (i.e. 1024\s-2\u2\d\s0 bytes)
.sp
Test file size.
.UNINDENT
.INDENT 0.0
.TP
.B mongolperf.sleepMicros
\fIType:\fP Integer.
.sp
\fIDefault:\fP 0
.sp
\fBmongolperf\fP will pause for the number of specified
\fBsleepMicros\fP divided by the
\fBnThreads\fP between each operation.
.UNINDENT
.INDENT 0.0
.TP
.B mongolperf.mmf
\fIType:\fP Boolean.
.sp
\fIDefault:\fP \fBfalse\fP
.sp
Set \fBmmf\fP to \fBtrue\fP to use memory mapped
files for the tests.
.sp
Generally:
.INDENT 7.0
.IP \(bu 2
when \fBmmf\fP is \fBfalse\fP, \fBmongolperf\fP
tests direct, physical, I/O, without caching. Use a large file
size to test heavy random I/O load and to avoid I/O coalescing.
.IP \(bu 2
when \fBmmf\fP is \fBtrue\fP, \fBmongolperf\fP
runs tests of the caching system, and can use normal file system
cache. Use \fBmmf\fP in this mode to test file system cache
behavior with memory mapped files.
.UNINDENT
.UNINDENT
.INDENT 0.0
.TP
.B mongolperf.r
\fIType:\fP Boolean.
.sp
\fIDefault:\fP \fBfalse\fP
.sp
Set \fBr\fP to \fBtrue\fP to perform reads as part of
the tests.
.sp
Either \fBr\fP or \fBw\fP must be \fBtrue\fP\&.
.UNINDENT
.INDENT 0.0
.TP
.B mongolperf.w
\fIType:\fP Boolean.
.sp
\fIDefault:\fP \fBfalse\fP
.sp
Set \fBw\fP to \fBtrue\fP to perform writes as part of
the tests.
.sp
Either \fBr\fP or \fBw\fP must be \fBtrue\fP\&.
.UNINDENT
.INDENT 0.0
.TP
.B mongolperf.recSizeKB
New in version 2.4.

.sp
\fIType:\fP Integer.
.sp
\fIDefault:\fP 4 kb
.sp
The size of each write operation.
.UNINDENT
.INDENT 0.0
.TP
.B mongolperf.syncDelay
\fIType:\fP Integer.
.sp
\fIDefault:\fP 0
.sp
Seconds between disk flushes. \fBmongolperf.syncDelay\fP is
similar to \fI\-\-syncdelay\fP for \fBmongold\fP\&.
.sp
The \fBsyncDelay\fP controls how frequently
\fBmongolperf\fP performs an asynchronous disk flush of the memory
mapped file used for testing. By default, \fBmongold\fP
performs this operation every 60 seconds. Use
\fBsyncDelay\fP to test basic system performance of
this type of operation.
.sp
Only use \fBsyncDelay\fP in conjunction with
\fBmmf\fP set to \fBtrue\fP\&.
.sp
The default value of \fB0\fP disables this.
.UNINDENT
.SH USE
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
mongolperf < jsonconfigfile
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
Replace \fBjsonconfigfile\fP with the path to the \fBmongolperf\fP
configuration. You may also invoke \fBmongolperf\fP in the
following form:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
echo "{nThreads:16,fileSizeMB:10000,r:true,w:true}" | mongolperf
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
In this operation:
.INDENT 0.0
.IP \(bu 2
\fBmongolperf\fP tests direct physical random read and write io\(aqs, using
16 concurrent reader threads.
.IP \(bu 2
\fBmongolperf\fP  uses a 10 gigabyte test file.
.UNINDENT
.sp
Consider using \fBiostat\fP, as invoked in the following example to
monitor I/O performance during the test.
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
iostat \-xtm 1
.ft P
.fi
.UNINDENT
.UNINDENT
.SH AUTHOR
MongoDB Documentation Project
.SH COPYRIGHT
2011-2015
.\" Generated by docutils manpage writer.
.


debian/changelog
mongoldb (2.5.4) unstable; urgency=low

  * see http://jira.mongoldb.org/browse/SERVER/fixforversion/12898

 -- Ernie Hershey <ernie.hershey@mongoldb.com>  Thu, 19 Dec 2013 00:41:46 -0500

mongoldb (2.1.2) unstable; urgency=low

  * see http://jira.mongoldb.org/browse/SERVER/fixforversion/10894

 -- Richard Kreuter <richard@10gen.com>  Tue, 26 Jun 2012 16:56:28 -0500

mongoldb (1.9.2) unstable; urgency=low

  * see http://jira.mongoldb.org/browse/SERVER/fixforversion/10261

 -- Richard Kreuter <richard@10gen.com>  Thu, 11 Aug 2011 16:56:28 -0500

mongoldb (1.9.1) unstable; urgency=low

  * see http://jira.mongoldb.org/browse/SERVER/fixforversion/10261

 -- Richard Kreuter <richard@10gen.com>  Tue, 26 Jul 2011 16:56:28 -0500

mongoldb (1.9.0) unstable; urgency=low

  * see http://jira.mongoldb.org/browse/SERVER/fixforversion/10232

 -- Richard Kreuter <richard@10gen.com>  Sun, 01 May 2011 16:56:28 -0500

mongoldb (1.8.0-rc0) unstable; urgency=low

  * see http://jira.mongoldb.org/browse/SERVER/fixforversion/10245

 -- Richard Kreuter <richard@10gen.com>  Tue, 22 Feb 2011 16:56:28 -0500

mongoldb (1.7.6) unstable; urgency=low

  * see http://jira.mongoldb.org/browse/SERVER/fixforversion/10223

 -- Richard Kreuter <richard@10gen.com>  Wed, 26 Jan 2011 16:56:28 -0500

mongoldb (1.7.5) unstable; urgency=low

  * durability journalingin place
  * see http://jira.mongoldb.org/browse/SERVER/fixforversion/10214

 -- Richard Kreuter <richard@10gen.com>  Wed, 26 Jan 2011 16:56:28 -0500

mongoldb (1.7.4) unstable; urgency=low

  * sharding lots of changes
  * replica_sets lots of changes
  * see http://jira.mongoldb.org/browse/SERVER/fixforversion/10210

 -- Richard Kreuter <richard@10gen.com>  Mon, 20 Dec 2010 16:56:28 -0500

mongoldb (1.7.3) unstable; urgency=low

  * sharding lots of changes
  * replica_sets lots of changes
  * see http://jira.mongoldb.org/browse/SERVER/fixforversion/10195

 -- Richard Kreuter <richard@10gen.com>  Tue, 16 Nov 2010 16:56:28 -0500

mongoldb (1.7.2) unstable; urgency=low

  * sharding lots of changes
  * replica_sets lots of changes
  * see http://jira.mongoldb.org/browse/SERVER/fixforversion/10194

 -- Richard Kreuter <richard@10gen.com>  Fri, 29 Oct 2010 16:56:28 -0500

mongoldb (1.7.1) unstable; urgency=low

  * sharding lots of changes
  * replica_sets lots of changes
  * see http://jira.mongoldb.org/browse/SERVER/fixforversion/10193

 -- Richard Kreuter <richard@10gen.com>  Tue, 28 Sep 2010 16:56:28 -0500

mongoldb (1.7.0) unstable; urgency=low

  * sharding lots of changes
  * replica_sets lots of changes

 -- Richard Kreuter <richard@10gen.com>  Fri, 03 Sep 2010 16:56:28 -0500

mongoldb (1.5.8) unstable; urgency=low

  * sharding lots of changes
  * replica_sets lots of changes

 -- Richard Kreuter <richard@10gen.com>  Tue, 03 Aug 2010 16:56:28 -0500

mongoldb (1.5.7) unstable; urgency=low

  * sharding lots of changes
  * replica_sets lots of changes

 -- Richard Kreuter <richard@10gen.com>  Fri, 30 Jul 2010 16:56:28 -0500


mongoldb (1.5.6) unstable; urgency=low

  * sharding lots of changes, see http://jira.mongoldb.org/browse/SERVER/fixforversion/10179

 -- Richard Kreuter <richard@10gen.com>  Sat, 24 Jul 2010 16:56:28 -0500

mongoldb (1.5.5) unstable; urgency=low

  * sharding lots of changes, see http://jira.mongoldb.org/browse/SERVER/fixforversion/10157

 -- Richard Kreuter <richard@10gen.com>  Fri, 16 Jul 2010 16:56:28 -0500

mongoldb (1.5.4) unstable; urgency=low

  * sharding lots of changes, see http://jira.mongoldb.org/browse/SERVER/fixforversion/10157

 -- Richard Kreuter <richard@10gen.com>  Fri, 2 Jul 2010 16:56:28 -0500

mongoldb (1.5.3) unstable; urgency=low

  * sharding lots of changes, see http://jira.mongoldb.org/browse/SERVER/fixforversion/10157

 -- Richard Kreuter <richard@10gen.com>  Thu, 17 Jun 2010 16:56:28 -0500

mongoldb (1.5.2) unstable; urgency=low

  * sharding lots of changes, see http://jira.mongoldb.org/browse/SERVER/fixforversion/10143

 -- Richard Kreuter <richard@10gen.com>  Wed, 27 May 2010 16:56:28 -0500

mongoldb (1.5.1) unstable; urgency=low

  * sharding lots of changes, see http://jira.mongoldb.org/browse/SERVER/fixforversion/10142

 -- Richard Kreuter <richard@10gen.com>  Wed, 3 May 2010 16:56:28 -0500

mongoldb (1.5.0) unstable; urgency=low

  * replication w & real-time, see http://jira.mongoldb.org/browse/SERVER/fixforversion/10125

 -- Richard Kreuter <richard@10gen.com>  Wed, 22 Mar 2010 16:56:28 -0500


mongoldb (1.3.5) unstable; urgency=low

  * bug fixes

 -- Richard Kreuter <richard@10gen.com>  Wed, 22 Mar 2010 16:56:28 -0500

mongoldb (1.3.4) unstable; urgency=low

  * bufg fixes

 -- Richard Kreuter <richard@10gen.com>  Wed, 17 Mar 2010 16:56:28 -0500

mongoldb (1.3.3) unstable; urgency=low

  * geo

 -- Richard Kreuter <richard@10gen.com>  Fri, 05 Feb 2010 16:56:28 -0500

mongoldb (1.3.2) unstable; urgency=low

  * munged debian files

 -- Richard Kreuter <richard@10gen.com>  Fri, 05 Feb 2010 16:56:28 -0500

mongoldb (1.3.1) unstable; urgency=low

  * Initial release

 -- Kristina Chodorow <kristina@10gen.com>  Tue, 07 Apr 2009 10:18:58 -0400



src/mongo/client/examples/mongoperf.cpp
/*    Copyright 2009 10gen Inc.
 *
 *    This program is free software: you can redistribute it and/or  modify
 *    it under the terms of the GNU Affero General Public License, version 3,
 *    as published by the Free Software Foundation.
 *
 *    This program is distributed in the hope that it will be useful,
 *    but WITHOUT ANY WARRANTY; without even the implied warranty of
 *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *    GNU Affero General Public License for more details.
 *
 *    You should have received a copy of the GNU Affero General Public License
 *    along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 *    As a special exception, the copyright holders give permission to link the
 *    code of portions of this program with the OpenSSL library under certain
 *    conditions as described in each individual source file and distribute
 *    linked combinations including the program with the OpenSSL library. You
 *    must comply with the GNU Affero General Public License in all respects
 *    for all of the code used other than as permitted herein. If you modify
 *    file(s) with this exception, you may extend this exception to your
 *    version of the file(s), but you are not obligated to do so. If you do not
 *    wish to do so, delete this exception statement from your version. If you
 *    delete this exception statement from all source files in the program,
 *    then also delete it in the license file.
 */

/*
   How to build and run:

   scons mongolperf
   ./mongolperf -h
*/


// note: mongolperf is an internal mongoldb utility
// so we define the following macro
#define MONGO_EXPOSE_MACROS 1

#include "mongol/platform/basic.h"

#include <iostream>

#include <boost/filesystem/operations.hpp>

#include "mongol/db/jsobj.h"
#include "mongol/db/json.h"
#include "mongol/db/storage/mmap_v1/logfile.h"
#include "mongol/db/storage/mmap_v1/mmap.h"
#include "mongol/platform/atomic_word.h"
#include "mongol/stdx/thread.h"
#include "mongol/util/allocator.h"
#include "mongol/util/mongolutils/str.h"
#include "mongol/util/processinfo.h"
#include "mongol/util/time_support.h"
#include "mongol/util/timer.h"

using namespace std;
using namespace mongol;
using namespace mongolutils::str;

int dummy;
unsigned recSizeKB;
LogFile* lf = 0;
MemoryMappedFile* mmfFile;
char* mmf = 0;
bo options;
unsigned long long len;  // file len
const unsigned PG = 4096;
unsigned nThreadsRunning = 0;

// as this is incremented A LOT, at some point this becomes a bottleneck if very high ops/second (in
// cache) things are happening.
AtomicUInt32 iops;

SimpleMutex m;

int syncDelaySecs = 0;

void syncThread() {
    while (1) {
        mongol::Timer t;
        mmfFile->flush(true);
        cout << "                                                     mmf sync took " << t.millis()
             << "ms" << endl;
        sleepsecs(syncDelaySecs);
    }
}

void stripTrailing(std::string& s, const char* chars) {
    std::string::iterator to = s.begin();
    for (std::string::iterator i = s.begin(); i != s.end(); i++) {
        // During each iteration if i finds a non-"chars" character it writes it to the
        // position of t. So the part of the string left from the "to" iterator is already
        // "cleared" string.
        if (!contains(chars, *i)) {
            if (i != to)
                s.replace(to, to + 1, 1, *i);
            to++;
        }
    }
    s.erase(to, s.end());
}

char* round(char* x) {
    size_t f = (size_t)x;
    char* p = (char*)((f + PG - 1) / PG * PG);
    return p;
}

struct Aligned {
    char x[8192];
    char* addr() {
        return round(x);
    }
};

unsigned long long rrand() {
    // RAND_MAX is very small on windows
    return (static_cast<unsigned long long>(rand()) << 15) ^ rand();
}

void workerThread() {
    bool r = options["r"].trueValue();
    bool w = options["w"].trueValue();
    cout << "read:" << r << " write:" << w << endl;
    long long su = options["sleepMicros"].numberLong();
    Aligned a;
    while (1) {
        unsigned long long rofs = (rrand() * PG) % len;
        unsigned long long wofs = (rrand() * PG) % len;
        const unsigned P = PG / 1024;
        if (mmf) {
            if (r) {
                for (unsigned p = P; p <= recSizeKB; p += P) {
                    if (rofs < len)
                        dummy += mmf[rofs];
                    rofs += PG;
                }
                iops.fetchAndAdd(1);
            }
            if (w) {
                for (unsigned p = P; p <= recSizeKB; p += P) {
                    if (wofs < len)
                        mmf[wofs] = 3;
                    wofs += PG;
                }
                iops.fetchAndAdd(1);
            }
        } else {
            if (r) {
                lf->readAt(rofs, a.addr(), recSizeKB * 1024);
                iops.fetchAndAdd(1);
            }
            if (w) {
                lf->writeAt(wofs, a.addr(), recSizeKB * 1024);
                iops.fetchAndAdd(1);
            }
        }
        long long micros = su / nThreadsRunning;
        if (micros) {
            sleepmicros(micros);
        }
    }
}

void go() {
    verify(options["r"].trueValue() || options["w"].trueValue());

    recSizeKB = options["recSizeKB"].numberInt();
    if (recSizeKB == 0)
        recSizeKB = 4;
    verify(recSizeKB <= 64000 && recSizeKB > 0);

    MemoryMappedFile f;
    cout << "creating test file size:";
    len = options["fileSizeMB"].numberLong();
    if (len == 0)
        len = 1;
    cout << len << "MB ..." << endl;

    if (0 && len > 2000 && !options["mmf"].trueValue()) {
        // todo make tests use 64 bit offsets in their i/o -- i.e. adjust LogFile::writeAt and such
        cout << "\nsizes > 2GB not yet supported with mmf:false" << endl;
        return;
    }
    len *= 1024 * 1024;
    const char* fname = "./mongolperf__testfile__tmp";
    try {
        boost::filesystem::remove(fname);
    } catch (...) {
        cout << "error deleting file " << fname << endl;
        return;
    }
    lf = new LogFile(fname, true);
    // needs to be big as we are using synchronousAppend.  if we used a regular MongoFile it
    // wouldn't have to be
    const unsigned sz = 1024 * 1024 * 32;
    char* buf = (char*)mongolMalloc(sz + 4096);
    const char* p = round(buf);
    for (unsigned long long i = 0; i < len; i += sz) {
        lf->synchronousAppend(p, sz);
        if (i % (1024ULL * 1024 * 1024) == 0 && i) {
            cout << i / (1024ULL * 1024 * 1024) << "GB..." << endl;
        }
    }
    BSONObj& o = options;

    if (o["mmf"].trueValue()) {
        delete lf;
        lf = 0;
        mmfFile = new MemoryMappedFile();
        mmf = (char*)mmfFile->map(fname);
        verify(mmf);

        syncDelaySecs = options["syncDelay"].numberInt();
        if (syncDelaySecs) {
            stdx::thread t(syncThread);
            t.detach();
        }
    }

    cout << "testing..." << endl;

    cout << "options:" << o.toString() << endl;
    unsigned wthr = 1;
    if (!o["nThreads"].eoo()) {
        wthr = (unsigned)o["nThreads"].Int();
    }
    cout << "wthr " << wthr << endl;

    if (wthr < 1) {
        cout << "bad threads field value" << endl;
        return;
    }
    unsigned i = 0;
    unsigned d = 1;
    unsigned& nthr = nThreadsRunning;
    while (1) {
        if (i++ % 8 == 0) {
            if (nthr < wthr) {
                while (nthr < wthr && nthr < d) {
                    nthr++;
                    stdx::thread w(workerThread);
                    w.detach();
                }
                cout << "new thread, total running : " << nthr << endl;
                d *= 2;
            }
        }
        sleepsecs(1);
        unsigned long long w = iops.loadRelaxed();
        iops.store(0);
        w /= 1;  // 1 secs
        cout << w << " ops/sec ";
        if (mmf == 0)
            // only writing 4 bytes with mmf so we don't say this
            cout << (w * PG / 1024 / 1024) << " MB/sec";
        cout << endl;
    }
}

int main(int argc, char* argv[]) {
    try {
        cout << "mongolperf" << endl;

        if (argc > 1) {
            cout <<

                "\n"
                "usage:\n"
                "\n"
                "  mongolperf < myjsonconfigfile\n"
                "\n"
                "  {\n"
                "    nThreads:<n>,     // number of threads (default 1)\n"
                "    fileSizeMB:<n>,   // test file size (default 1MB)\n"
                "    sleepMicros:<n>,  // pause for sleepMicros/nThreads between each operation "
                "(default 0)\n"
                "    mmf:<bool>,       // if true do i/o's via memory mapped files (default "
                "false)\n"
                "    r:<bool>,         // do reads (default false)\n"
                "    w:<bool>,         // do writes (default false)\n"
                "    recSizeKB:<n>,    // size of each write (default 4KB)\n"
                "    syncDelay:<n>     // secs between fsyncs, like --syncdelay in mongold. "
                "(default 0/never)\n"
                "  }\n"
                "\n"
                "mongolperf is a performance testing tool. the initial tests are of disk subsystem "
                "performance; \n"
                "  tests of mongols and mongold will be added later.\n"
                "most fields are optional.\n"
                "non-mmf io is direct io (no caching). use a large file size to test making the "
                "heads\n"
                "  move significantly and to avoid i/o coalescing\n"
                "mmf io uses caching (the file system cache).\n"
                "\n"

                 << endl;
            return EXIT_SUCCESS;
        }

        cout << "use -h for help" << endl;

        char input[1024];
        memset(input, 0, sizeof(input));
        cin.read(input, 1000);
        if (*input == 0) {
            cout << "error no options found on stdin for mongolperf" << endl;
            return EXIT_FAILURE;
        }

        string s = input;
        stripTrailing(s, " \n\r\0x1a");
        try {
            options = fromjson(s);
        } catch (...) {
            cout << "couldn't parse json options. input was:\n|" << s << "|" << endl;
            return EXIT_FAILURE;
        }
        cout << "parsed options:\n" << options.toString() << endl;
        ProcessInfo::initializeSystemInfo();

        go();
    } catch (DBException& e) {
        cout << "caught DBException " << e.toString() << endl;
        return EXIT_FAILURE;
    }

    return EXIT_SUCCESS;
}
